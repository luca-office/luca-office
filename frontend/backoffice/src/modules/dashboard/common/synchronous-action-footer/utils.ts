import {IconName, TimeUnit} from "shared/enums"
import {ProjectModule} from "shared/models"
import {secondsToGivenTimeUnit} from "shared/utils"

export const getSecondaryButtonIcon = (
  hasSurveyStarted: boolean,
  hasSurveyEnded: boolean,
  hasNextModule: boolean,
  isAsyncSurvey: boolean
) => {
  if (hasSurveyStarted && !hasSurveyEnded) {
    return hasNextModule && !isAsyncSurvey ? IconName.Play : IconName.Check
  } else if (hasSurveyEnded) {
    return undefined
  } else {
    return IconName.Play
  }
}

export const getModuleDurationInMinutes = ({scenario}: ProjectModule) => {
  if (scenario && scenario.maxDurationInSeconds !== null) {
    return secondsToGivenTimeUnit(TimeUnit.Minute, scenario.maxDurationInSeconds)
  } else {
    return null
  }
}
