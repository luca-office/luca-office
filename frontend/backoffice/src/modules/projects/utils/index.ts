export * from "./sorting"
export * from "./state"
