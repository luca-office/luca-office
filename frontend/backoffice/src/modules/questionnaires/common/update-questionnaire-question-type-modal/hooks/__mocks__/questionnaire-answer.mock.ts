import {QuestionnaireAnswer} from "shared/models"

export const questionnaireAnswerMock = (
  questionId: UUID = "18316dac-2355-4e45-a7f3-f339461a488f"
): QuestionnaireAnswer => ({
  __typename: "QuestionnaireAnswer",
  createdAt: new Date(2020, 10, 5).toISOString(),
  modifiedAt: new Date(2020, 10, 15).toISOString(),
  id: "1fce8289-71fa-4568-bccc-329dc4cc67eb",
  text: "answer",
  questionId,
  isCorrect: false,
  position: 1
})
