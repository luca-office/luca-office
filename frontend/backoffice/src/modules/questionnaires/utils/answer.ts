export const getScore = (scoreValue: string, lowerLimit = 0, upperLimit = 99): number => {
  const score = parseInt(scoreValue, 10) || lowerLimit
  return score < lowerLimit ? lowerLimit : score > upperLimit ? upperLimit : score
}
