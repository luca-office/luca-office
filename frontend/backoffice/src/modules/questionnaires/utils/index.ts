export * from "./answer"
export * from "./questions"
