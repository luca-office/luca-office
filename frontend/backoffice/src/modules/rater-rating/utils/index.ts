export * from "./coding-dimension"
export * from "./participant"
export * from "./progress-indicator-status"
export * from "./questions"
export * from "./rating-status"
