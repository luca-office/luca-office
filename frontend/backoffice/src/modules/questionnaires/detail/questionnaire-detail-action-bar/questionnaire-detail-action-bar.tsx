import {css} from "@emotion/react"
import * as React from "react"
import {Button, CardFooterItem} from "shared/components"
import {IconName} from "shared/enums"
import {Questionnaire} from "shared/models"
import {spacingLarge} from "shared/styles"
import {WithLucaTranslation} from "shared/translations"
import {formatDateFromString, Option} from "shared/utils"

export interface QuestionnaireDetailActionBarProps extends WithLucaTranslation {
  readonly questionnaire: Option<Questionnaire>
  readonly openPreview: () => void
}

export const QuestionnaireDetailActionBar: React.FC<QuestionnaireDetailActionBarProps> = ({
  openPreview,
  questionnaire: questionnaireOption,
  t
}) =>
  questionnaireOption
    .map(questionnaire => (
      <React.Fragment>
        <CardFooterItem
          title={t("detail_card__title_createdAt")}
          icon={IconName.Calendar}
          text={formatDateFromString(questionnaire.createdAt)}
        />
        <CardFooterItem
          customStyles={styles.author}
          title={t("detail_card__title_author")}
          icon={IconName.Profile}
          text={`${questionnaire.author.firstName} ${questionnaire.author.lastName}`}
        />
        <Button onClick={openPreview} customStyles={styles.preview}>
          {t("preview")}
        </Button>
      </React.Fragment>
    ))
    .orNull()

const styles = {
  author: css({
    marginLeft: spacingLarge
  }),
  preview: css({
    marginLeft: "auto",
    justifySelf: "flex-end"
  })
}
