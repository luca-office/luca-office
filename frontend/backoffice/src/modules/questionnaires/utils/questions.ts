import {getQuestionTypeIconName} from "shared/components"
import {QuestionScoringType, QuestionType} from "shared/graphql/generated/globalTypes"
import {FreetextQuestionCodingCriterion, QuestionnaireAnswer, QuestionnaireQuestion} from "shared/models"
import {exists, Option} from "shared/utils"
import {ResortableEntity} from "../../../models"

export const convertQuestionsToResortableEntity = (
  questionsOption: Option<QuestionnaireQuestion[]>
): Option<ResortableEntity[]> =>
  questionsOption.map<ResortableEntity[]>(questions =>
    questions.map(question => ({
      id: question.id,
      position: question.position,
      icon: getQuestionTypeIconName(question.questionType),
      title: question.text
    }))
  )

export const convertAnswersToResortableEntity = (
  answersOption: Option<QuestionnaireAnswer[]>
): Option<ResortableEntity[]> =>
  answersOption.map<ResortableEntity[]>(answers =>
    answers.map(answer => ({
      id: answer.id,
      position: answer.position,
      title: answer.text
    }))
  )

/**
 * if the user changes the question type from a type that support a certain scoring to
 * a type that does not support scoring, we need to reset the scoring
 * @param isScoringEnabled - does the current question have a scoring setting !== None
 * @param sourceQuestionType - type of current question
 * @param targetQuestionType - selected type that should be switched to
 */
export const isScoringResetNecessary = (
  isScoringEnabled: boolean,
  sourceQuestionType: QuestionType,
  targetQuestionType: QuestionType
) => {
  return targetQuestionType !== QuestionType.FreeText && sourceQuestionType !== targetQuestionType && isScoringEnabled
}

const hasEmptyQuestionTitle = (questions: QuestionnaireQuestion[]) =>
  questions.some(question => question.text.trim() === "")

const hasEmptyQuestionAnswers = (questions: QuestionnaireQuestion[]) =>
  questions.some(question => question.answers.some(answer => answer.text.trim() === ""))

const isNoScoreFreetextCriterion = (criterion: FreetextQuestionCodingCriterion) =>
  criterion.description.trim() === "" && criterion.score === 0

const hasEmptyCodingCriteria = (questions: QuestionnaireQuestion[]) =>
  questions
    .filter(question => question.questionType === QuestionType.FreeText)
    .some(freeTextQuestion => {
      // only allow extact one criterion with empty description and score of 0
      return freeTextQuestion.freetextQuestionCodingCriteria.filter(isNoScoreFreetextCriterion).length > 1
        ? true
        : freeTextQuestion.freetextQuestionCodingCriteria
            .filter(criterion => !isNoScoreFreetextCriterion(criterion))
            .some(criterion => criterion.description.trim() === "") &&
            freeTextQuestion.scoringType !== QuestionScoringType.None
    })

export const containsEmptyTitleAnswersOrCriteria = (questions: QuestionnaireQuestion[]) =>
  hasEmptyQuestionTitle(questions) || hasEmptyQuestionAnswers(questions) || hasEmptyCodingCriteria(questions)

export const containsUnsetCorrectAnswer = (questions: QuestionnaireQuestion[]): boolean =>
  exists(question => {
    const needsCorrectAnswer =
      (question.questionType === QuestionType.SingleChoice || question.questionType === QuestionType.MultipleChoice) &&
      question.scoringType !== QuestionScoringType.None

    return needsCorrectAnswer && !exists(answer => answer.isCorrect, question.answers)
  }, questions)
