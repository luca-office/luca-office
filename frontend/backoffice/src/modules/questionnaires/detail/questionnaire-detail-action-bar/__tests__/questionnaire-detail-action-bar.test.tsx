import {shallow} from "enzyme"
import * as React from "react"
import {create} from "react-test-renderer"
import {Button, CardFooterItem} from "shared/components"
import {questionnairesMock} from "shared/graphql/__mocks__"
import {formatDateFromString, Option} from "shared/utils"
import {fakeTranslate} from "sharedTests/utils/translate-mock"
import {QuestionnaireDetailActionBar, QuestionnaireDetailActionBarProps} from "../questionnaire-detail-action-bar"

const previewSpy = jest.fn()
const questionnaire = questionnairesMock[0]
const defaultProps: QuestionnaireDetailActionBarProps = {
  t: fakeTranslate,
  openPreview: previewSpy,
  questionnaire: Option.of(questionnaire)
}
const getComponent = (props?: Partial<QuestionnaireDetailActionBarProps>) => (
  <QuestionnaireDetailActionBar {...defaultProps} {...props} />
)

describe("questionnaire-detail-action-bar", () => {
  it("renders correctly", () => {
    const component = create(getComponent())
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
  it("has correct structure", () => {
    const tree = shallow(getComponent())

    expect(tree.find(CardFooterItem)).toHaveLength(2)
    expect(tree.find(CardFooterItem).first().html()).toContain(formatDateFromString(questionnaire.createdAt))
    expect(tree.find(CardFooterItem).last().html()).toContain(questionnaire.author.firstName)
    expect(tree.find(CardFooterItem).last().html()).toContain(questionnaire.author.lastName)
    expect(tree.find(Button)).toHaveLength(1)
  })
  it("executes actions correctly", () => {
    const tree = shallow(getComponent())

    tree.find(Button).simulate("click")
    expect(previewSpy).toHaveBeenCalled()
  })
})
