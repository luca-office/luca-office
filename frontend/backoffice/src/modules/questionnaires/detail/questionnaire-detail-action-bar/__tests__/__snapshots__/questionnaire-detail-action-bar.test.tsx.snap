// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`questionnaire-detail-action-bar renders correctly 1`] = `
[
  <div
    className="css-1839z0z-CardFooterItem"
    title="detail_card__title_createdAt"
  >
    <div
      className="css-1lrzij5-Icon"
      dangerouslySetInnerHTML={
        {
          "__html": {},
        }
      }
    />
    <div
      className="css-1de4iva-Text"
    >
      21.07.2020
    </div>
  </div>,
  <div
    className="css-nsmsx2-CardFooterItem"
    title="detail_card__title_author"
  >
    <div
      className="css-1lrzij5-Icon"
      dangerouslySetInnerHTML={
        {
          "__html": {},
        }
      }
    />
    <div
      className="css-1de4iva-Text"
    >
      John Doe
    </div>
  </div>,
  <button
    className="css-tmvgiy-Button"
    disabled={false}
    onClick={[Function]}
  >
    <div
      className="css-95j2hm-Button"
    >
      preview
    </div>
  </button>,
]
`;
